src/crates.rs

```diff
@@ -0,0 +1,273 @@
+    /// and if succesful, return a handle to it.
+    pub fn download_new(config: &Config) -> Result<CrateFile, Error> {
+        let pkgname = &config.package.name;
+        let pkgver = &config.package.version;
+        // This is downloaded to a temporary directory, instead of the current
+        // directory, to avoid the issue where cargo thinks its in a workspace.
+        // https://github.com/rust-lang/cargo/issues/5418
+        let tempdir_handle = tempfile::tempdir()?;
+        let crate_file_prefix = format!("{pkgname}-{pkgver}");
+        let crate_file_extension = "tar.gz";
+        let crate_file_path_full = tempdir_handle
+            .as_ref()
+            .join(&crate_file_prefix)
+            .with_extension(crate_file_extension);
+        let crate_url =
+            format!("https://static.crates.io/crates/{pkgname}/{pkgname}-{pkgver}.crate");
+        let success = Command::new("curl")
+            .arg("--output")
+            .arg(crate_file_path_full)
+            .arg(&crate_url)
+            .status()?
+            .success();
+        if !success {
+            return Err(Error::DownloadingCrate { crate_url });
+        }
+        Ok(CrateFile {
+            tempdir_handle,
+            crate_file_prefix,
+            crate_file_extension,
+            config,
+        })
+    }
+    /// Get the sha256sum for the downloaded .crate file.
+    // NOTE: Possibly future refactor target is crate::sha256sum
+    pub fn get_sha256sum(&self) -> Result<String, Error> {
+        let crate_file_path = self
+            .tempdir_handle
+            .as_ref()
+            .join(&self.crate_file_prefix)
+            .with_extension(self.crate_file_extension);
+        let bytes = std::fs::read(crate_file_path)?;
+        let digest = Hash::hash(&bytes);
+        let hex = digest.iter().fold(String::new(), |mut output, b| {
+            write!(output, "{:02x}", b).expect("Write to a string should not fail");
+            output
+        });
+        Ok(hex)
+    }
+    /// Check to see if the license needs to be installed.
+    /// If so, extract the crate, and get the relative path of the LICENSE file
+    /// So that we know what 'install' command to run in the PKGBUILD.
+    pub fn get_license(&self) -> Result<Option<DirEntry>, Error> {
+        crate::alert_if_must_copy_license(&self.config.package.license)
+            .then(|| self.license_file())
+            .transpose()
+    }
+    /// Extract the crate, and get the relative path of the LICENSE file
+    fn license_file(&self) -> Result<DirEntry, Error> {
+        let crate_filename =
+            PathBuf::from(&self.crate_file_prefix).with_extension(self.crate_file_extension);
+        if !Command::new("tar")
+            .current_dir(self.tempdir_handle.as_ref())
+            .arg("-xvzf")
+            .arg(&crate_filename)
+            .status()?
+            .success()
+        {
+            return Err(Error::ExtractingCrate { crate_filename });
+        };
```

Comment:
This logic appears to be shared, so let's extract it out into a function.

Reply:
actioned